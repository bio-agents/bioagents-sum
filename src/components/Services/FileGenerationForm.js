import React from 'react'
import { Field, formValueSelector, reduxForm } from 'redux-form'
import buildActionCreators from '../../helpers/buildActionCreators'
import * as ActionTypes from '../../constants/actionTypes'
import { connect } from 'react-redux'
import { reportType } from '../../constants/generateFile'
import SelectField from '../common/SelectField'
import { Button, Col, Form, FormGroup, Panel, ToggleButton } from 'react-bootstrap'
import ToggleRadioButtonGroup from '../common/ToggleRadioButtonGroup'
import TextField from '../common/TextField'
import ToggleCheckboxButtonGroup from '../common/ToggleCheckboxButtonGroup'
import Checkbox from '../common/Checkbox'

class FileGenerationForm extends React.PureComponent {
  generateFile = event => {
    event.preventDefault()
    const { generateFile, list } = this.props
    generateFile({ list })
  }

  render () {
    const { reset, submitting, reportTypeChosen, sortByChosen, orderChosen, includePropsChosen } = this.props

    return (
      <Panel bsStyle='warning'>
        <Form horizontal onSubmit={this.generateFile}>
          <Field name='reportType' component={SelectField} label='Choose type of report' >
            <option value={reportType.DOCX}>{'DOCX agents report'}</option>
            <option value={reportType.XLSX}>{'XLSX agents report'}</option>
            <option value={reportType.CHART}>{'Citations chart'}</option>
          </Field>
          {reportTypeChosen === reportType.CHART
            ? <Field name='createGraph' component={Checkbox} label='I want to create my own graph by selecting agents' />
            : <div>
              <Field name='sortBy' component={ToggleRadioButtonGroup} valueChosen={sortByChosen} label='Sort by' >
                <ToggleButton value='citations'>{'Citations'}</ToggleButton>
                <ToggleButton value='name'>{'Name'}</ToggleButton>
              </Field>
              <Field name='order' component={ToggleRadioButtonGroup} valueChosen={orderChosen} label='Order' >
                <ToggleButton value='descend'>{'Descending'}</ToggleButton>
                <ToggleButton value='ascend'>{'Ascending'}</ToggleButton>
              </Field>
              <Field name='takeFirstX' component={TextField} label='Number of agents' />
              <Field name='includeProps' component={ToggleCheckboxButtonGroup} valueChosen={includePropsChosen} label='Include'>
                <ToggleButton value='agentType'>{'Agent type'}</ToggleButton>
                <ToggleButton value='institute'>{'Institute'}</ToggleButton>
                <ToggleButton value='description'>{'Description'}</ToggleButton>
                <ToggleButton value='publication'>{'Publications'}</ToggleButton>
                <ToggleButton value='citations'>{'Citations'}</ToggleButton>
                <ToggleButton value='topic'>{'Topic'}</ToggleButton>
                <ToggleButton value='function'>{'Function'}</ToggleButton>
                <ToggleButton value='maturity'>{'Maturity'}</ToggleButton>
                <ToggleButton value='platform'>{'Platform'}</ToggleButton>
              </Field>
              <FormGroup>
                <Col smOffset={2} sm={10}>
                  <Button type='submit' disabled={submitting}>
                    {'Generate'}
                  </Button>
                  <Button type='button' disabled={submitting} onClick={reset}>
                    {'Clear Values'}
                  </Button>
                </Col>
              </FormGroup>
            </div>
          }
        </Form>
      </Panel>
    )
  }
}

FileGenerationForm = reduxForm({
  form: 'fileGenerationForm',
})(FileGenerationForm)

export default connect(state => {
  const selector = formValueSelector('fileGenerationForm')

  return {
    reportTypeChosen: selector(state, 'reportType'),
    sortByChosen: selector(state, 'sortBy'),
    orderChosen: selector(state, 'order'),
    includePropsChosen: selector(state, 'includeProps'),
    initialValues: { reportType: 'docx', sortBy: 'citations', order: 'descend', includeProps: [] },
  }
}, buildActionCreators({
  generateFile: ActionTypes.GENERATE_FILE,
}))(FileGenerationForm)
