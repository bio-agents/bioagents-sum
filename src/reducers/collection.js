import * as R from 'ramda'
import * as Actions from '../constants/actionTypes'
import { configCollection } from '../bioagentsSum/common/helperFunctions'

const initialState = {
  activeCollection: '',
  userEnteredCollection: false,
}

export const collection = (state = initialState, action) => {
  const { type, payload } = action

  switch (type) {
    case Actions.SET_COLLECTION:
      return R.evolve({
        activeCollection: R.always(payload.collection !== '' ? payload.collection : configCollection),
        userEnteredCollection: R.always(payload.collection !== '' ? payload.userEnteredCollection : false),
      })(state)

    default:
      return state
  }
}
