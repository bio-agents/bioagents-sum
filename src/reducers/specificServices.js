import * as R from 'ramda'
import * as Actions from '../constants/actionTypes'
import { pickData } from '../bioagentsSum/services/index'

const initialState = {
  count: 0,
  list: [],
  persistExpiresAt: '',
  serviceLoading: false,
  serviceLoaded: false,
  citationsLoading: false,
  citationsLoaded: false,
}

export const specificServicesWithName = (servicesName = '') =>
  (state = initialState, action) => {
    const { type, payload } = action

    if (payload && payload.name !== servicesName) return state

    switch (type) {
      case Actions.SERVICES_FETCH:
        return R.evolve({
          serviceLoading: R.T,
          serviceLoaded: R.F,
          citationsLoading: R.F,
          citationsLoaded: R.F,
        })(state)

      case Actions.SERVICES_FETCH_SUCCESS: {
        const { count, list } = payload.service
        let finalList = pickData(list)

        if (state.list.length !== 0) {
          const oldListByIds = R.groupBy(R.prop('id'), state.list)
          finalList = R.map(agent => R.mergeDeepLeft(agent, oldListByIds[agent.id] ? oldListByIds[agent.id][0] : {}), finalList)
        }

        return R.evolve({
          count: R.always(count),
          list: R.always(finalList),
          serviceLoading: R.F,
          serviceLoaded: R.T,
          citationsLoading: R.T,
          persistExpiresAt: R.always(Date.now() + 600000), // 86400000 ms == 24 hours
        })(state)
      }

      case Actions.SERVICES_FETCH_FAILURE: {
        return R.evolve({
          serviceLoading: R.F,
          serviceLoaded: R.F,
        })(state)
      }

      case Actions.CITATIONS_FETCH_SUCCESS: {
        const { count, list } = payload.service
        const pickedData = pickData(list)

        return R.evolve({
          count: R.always(count),
          list: R.always(pickedData),
          citationsLoading: R.F,
          citationsLoaded: R.T,
        })(state)
      }

      case Actions.CITATIONS_FETCH_FAILURE:
        return R.evolve({
          citationsLoading: R.F,
          citationsLoaded: R.F,
        })(state)

      case Actions.UPTIME_FETCH_SUCCESS: {
        const { count, list } = payload.service
        const pickedData = pickData(list)

        return R.evolve({
          count: R.always(count),
          list: R.always(pickedData),
        })(state)
      }

      case Actions.UPTIME_FETCH_FAILURE: {
        // todo
      }

      default:
        return state
    }
  }
