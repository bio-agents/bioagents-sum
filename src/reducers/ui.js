import * as R from 'ramda'
import * as Actions from '../constants/actionTypes'

const initialState = {
  showReportPage: false,
}

export const ui = (state = initialState, action) => {
  const { type } = action

  switch (type) {
    case Actions.REPORT_PAGE_CHANGE_VISIBILITY:
      return R.evolve({
        showReportPage: R.not,
      })(state)

    default:
      return state
  }
}
