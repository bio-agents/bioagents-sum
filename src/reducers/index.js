import { combineReducers } from 'redux'
import { routerReducer } from 'react-router-redux'
import { reducer as reduxFormReducer } from 'redux-form'
import { specificServicesWithName } from './specificServices'
import * as R from 'ramda'
import { ui } from './ui'
import { collection } from './collection'
import { getServicesNames } from '../bioagentsSum/common/helperFunctions'

const bioagentsSumApp = () => {
  const reducers = R.map(specificServicesWithName, getServicesNames)
  const servicesReducers = R.zipObj(getServicesNames, reducers)

  return combineReducers(R.merge(servicesReducers, {
    ui,
    collection,
    router: routerReducer,
    form: reduxFormReducer,
  }))
}

export default bioagentsSumApp
