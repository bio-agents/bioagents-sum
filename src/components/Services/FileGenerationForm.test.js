import React from 'react'
import renderer from 'react-test-renderer'
import { ConnectedRouter } from 'react-router-redux'
import { createStore } from 'redux'
import createHistory from 'history/createBrowserHistory'
import reducers from '../../reducers'
import { Provider } from 'react-redux'
import FileGenerationForm from './FileGenerationForm'

const store = createStore(reducers(), {})
const history = createHistory()

it('Renders MatrixCellWithLink with link and name correctly', () => {
  const component = renderer.create(
    <Provider store={store}>
      <ConnectedRouter history={history}>
        <FileGenerationForm />
      </ConnectedRouter>
    </Provider>
  )

  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
