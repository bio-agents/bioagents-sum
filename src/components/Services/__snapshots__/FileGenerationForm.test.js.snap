exports[`test Renders MatrixCellWithLink with link and name correctly 1`] = `
<div
  className="panel panel-warning"
  id={undefined}>
  <div
    className="panel-body">
    <form
      className="form-horizontal"
      onSubmit={[Function]}>
      <div
        className="form-group">
        <label
          className="col-sm-2 control-label"
          htmlFor="reportType">
          Choose type of report
        </label>
        <div
          className="col-sm-10">
          <select
            className="form-control"
            id="reportType"
            name="reportType"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            type={undefined}
            value="docx">
            <option
              value="docx">
              DOCX agents report
            </option>
            <option
              value="xlsx">
              XLSX agents report
            </option>
            <option
              value="chart">
              Citations chart
            </option>
          </select>
        </div>
      </div>
      <div>
        <div
          className="form-group">
          <label
            className="col-sm-2 control-label"
            htmlFor={undefined}>
            Sort by
          </label>
          <div
            className="col-sm-10">
            <div
              className="btn-group"
              data-toggle="buttons">
              <label
                className="btn btn-default active"
                disabled={false}
                type="button">
                <input
                  autoComplete="off"
                  checked={true}
                  name="sortBy"
                  onChange={[Function]}
                  type="radio"
                  value="citations" />
                Citations
              </label>
              <label
                className="btn btn-default"
                disabled={false}
                type="button">
                <input
                  autoComplete="off"
                  checked={false}
                  name="sortBy"
                  onChange={[Function]}
                  type="radio"
                  value="name" />
                Name
              </label>
            </div>
          </div>
        </div>
        <div
          className="form-group">
          <label
            className="col-sm-2 control-label"
            htmlFor={undefined}>
            Order
          </label>
          <div
            className="col-sm-10">
            <div
              className="btn-group"
              data-toggle="buttons">
              <label
                className="btn btn-default active"
                disabled={false}
                type="button">
                <input
                  autoComplete="off"
                  checked={true}
                  name="order"
                  onChange={[Function]}
                  type="radio"
                  value="descend" />
                Descending
              </label>
              <label
                className="btn btn-default"
                disabled={false}
                type="button">
                <input
                  autoComplete="off"
                  checked={false}
                  name="order"
                  onChange={[Function]}
                  type="radio"
                  value="ascend" />
                Ascending
              </label>
            </div>
          </div>
        </div>
        <div
          className="form-group">
          <label
            className="col-sm-2 control-label"
            htmlFor="takeFirstX">
            Number of agents
          </label>
          <div
            className="col-sm-10">
            <input
              className="form-control"
              id="takeFirstX"
              name="takeFirstX"
              onBlur={[Function]}
              onChange={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              onFocus={[Function]}
              type={undefined}
              value="" />
          </div>
        </div>
        <div
          className="form-group">
          <label
            className="col-sm-2 control-label"
            htmlFor={undefined}>
            Include
          </label>
          <div
            className="col-sm-10">
            <div
              className="btn-group"
              data-toggle="buttons">
              <label
                className="btn btn-default"
                disabled={false}
                type="button">
                <input
                  autoComplete="off"
                  checked={false}
                  name="includeProps"
                  onChange={[Function]}
                  type="checkbox"
                  value="agentType" />
                Agent type
              </label>
              <label
                className="btn btn-default"
                disabled={false}
                type="button">
                <input
                  autoComplete="off"
                  checked={false}
                  name="includeProps"
                  onChange={[Function]}
                  type="checkbox"
                  value="institute" />
                Institute
              </label>
              <label
                className="btn btn-default"
                disabled={false}
                type="button">
                <input
                  autoComplete="off"
                  checked={false}
                  name="includeProps"
                  onChange={[Function]}
                  type="checkbox"
                  value="description" />
                Description
              </label>
              <label
                className="btn btn-default"
                disabled={false}
                type="button">
                <input
                  autoComplete="off"
                  checked={false}
                  name="includeProps"
                  onChange={[Function]}
                  type="checkbox"
                  value="publication" />
                Publications
              </label>
              <label
                className="btn btn-default"
                disabled={false}
                type="button">
                <input
                  autoComplete="off"
                  checked={false}
                  name="includeProps"
                  onChange={[Function]}
                  type="checkbox"
                  value="citations" />
                Citations
              </label>
              <label
                className="btn btn-default"
                disabled={false}
                type="button">
                <input
                  autoComplete="off"
                  checked={false}
                  name="includeProps"
                  onChange={[Function]}
                  type="checkbox"
                  value="topic" />
                Topic
              </label>
              <label
                className="btn btn-default"
                disabled={false}
                type="button">
                <input
                  autoComplete="off"
                  checked={false}
                  name="includeProps"
                  onChange={[Function]}
                  type="checkbox"
                  value="function" />
                Function
              </label>
              <label
                className="btn btn-default"
                disabled={false}
                type="button">
                <input
                  autoComplete="off"
                  checked={false}
                  name="includeProps"
                  onChange={[Function]}
                  type="checkbox"
                  value="maturity" />
                Maturity
              </label>
              <label
                className="btn btn-default"
                disabled={false}
                type="button">
                <input
                  autoComplete="off"
                  checked={false}
                  name="includeProps"
                  onChange={[Function]}
                  type="checkbox"
                  value="platform" />
                Platform
              </label>
            </div>
          </div>
        </div>
        <div
          className="form-group">
          <div
            className="col-sm-10 col-sm-offset-2">
            <button
              className="btn btn-default"
              disabled={false}
              type="submit">
              Generate
            </button>
            <button
              className="btn btn-default"
              disabled={false}
              onClick={[Function]}
              type="button">
              Clear Values
            </button>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
